resource/resource.go

```diff
@@ -0,0 +1,49 @@
+// Package resource defines an interface for recording changes to InfluxDB resources.
+//
+// A resource is an entity in our system, e.g. an organization, task or bucket.
+// A change includes the creation, update or deletion of a resource.
+package resource
+
+import (
+	"time"
+
+	"github.com/influxdata/influxdb"
+)
+
+// Logger records changes to resources.
+type Logger interface {
+	// Log a change to a resource.
+	Log(Change) error
+}
+
+// Change to a resource.
+type Change struct {
+	// Type of change.
+	Type ChangeType
+	// ResourceID of the changed resource.
+	ResourceID influxdb.ID
+	// ResourceType that was changed.
+	ResourceType influxdb.ResourceType
+	// OrganizationID of the organization owning the changed resource.
+	OrganizationID influxdb.ID
+	// UserID of the user changing the resource.
+	UserID influxdb.ID
+	// ResourceBody after the change.
+	ResourceBody []byte
+	// Time when the resource was changed.
+	Time time.Time
+}
```

Comment:
Wanted to state the design here, for my own clarification and thinking aloud. tl;dr is this approach is sound 👍 

----

To know the specific changes made to a resource over time, we read its entire history and look at each prior `ResourceBody` to determine the deltas. Knowing that we will store the change history in InfluxDB, this approach is more efficient. Why? Prior values will be in the same column as the `ResourceBody` field and very likely stored in the same TSM block, so will be decompressed together. 💯 

----

For posterity, the alternative would be to include the `PreviousResourceBody` as an additional value as a separate column. Whilst this approach may be better for a row-based data store, it will require separate reads for a columnar-based store like InfluxDB. 

💯